File: Java/JavaComplexity/app/src/main/java/com/dodemy/javacomplexity/MainActivity.java
package com.dodemy.javacomplexity;

import android.os.Bundle;

import com.google.android.material.floatingactionbutton.FloatingActionButton;

import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.widget.Toolbar;

import android.view.View;
import android.view.Menu;
import android.view.MenuItem;
import android.widget.AdapterView;
import android.widget.ArrayAdapter;
import android.widget.AutoCompleteTextView;
import android.widget.EditText;
import android.widget.Spinner;
import android.widget.Toast;

import java.util.ArrayList;
import java.util.List;

import com.dodemy.javacomplexity.dao.IFoodDAO;
import com.dodemy.javacomplexity.dao.RetrofitClientInstance;
import com.dodemy.javacomplexity.dto.Food;
import com.dodemy.javacomplexity.dto.FoodType;
import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

public class MainActivity extends AppCompatActivity {

    private AutoCompleteTextView actType;
    private EditText edtCalories;
    private EditText edtCost;
    private EditText edtName;
    private EditText edtPrepTime;
    private Spinner spnFood;
    private Food food;
    private List<Food> allFoods;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        Toolbar toolbar = findViewById(R.id.toolbar);
        setSupportActionBar(toolbar);

        edtCalories = findViewById(R.id.edtCalories);
        edtCost = findViewById(R.id.edtCost);
        edtName = findViewById(R.id.edtName);
        edtPrepTime = findViewById(R.id.edtPrepTime);
        spnFood = findViewById(R.id.spnFood);
        actType = findViewById(R.id.actType);

        food = new Food();
        allFoods = new ArrayList<Food>();

        FloatingActionButton fab = findViewById(R.id.fab);
        fab.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                newFood();
            }
        });

        actType.setOnItemClickListener(new AdapterView.OnItemClickListener() {
            @Override
            public void onItemClick(AdapterView<?> parent, View view, int position, long id) {
                FoodType type = (FoodType) parent.getItemAtPosition(position);
                Toast.makeText(MainActivity.this, type + " Very nutritious!", Toast.LENGTH_LONG).show();
            }
        });
        spnFood.setOnItemSelectedListener(new AdapterView.OnItemSelectedListener() {
            @Override
            public void onItemSelected(AdapterView<?> parent, View view, int position, long id) {
                food = (Food) parent.getItemAtPosition(position);
                actType.setText(food.getType());
                String textCalories = Integer.toString(food.getCalories());
                edtCalories.setText(textCalories);
                edtCost.setText(food.getCost());
                String properName = properCase(food.getName());
                edtName.setText(properName);
                edtPrepTime.setText(food.getPrepTime());
            }

            @Override
            public void onNothingSelected(AdapterView<?> parent) {

            }
        });

        prepopulateFoods();
        updateSpinner();
        fetchFoodTypes();

    }

    private String properCase(String name) {
        String formatted = "";
        if (name.length() > 0) {
          formatted = name.substring(0,1).toUpperCase() + name.substring(1).toLowerCase();
        }
        return formatted;
    }

    private void prepopulateFoods() {
        Food mustard = new Food("Mustard", 0, "1.29", "0");
        allFoods.add(mustard);
        Food pickles = new Food("Pickles");
        allFoods.add(pickles);
        Food pawpaw = new Food("Paw Paw", 100, "10.00");
        allFoods.add(pawpaw);
        Food fujiApple = new Food();
        fujiApple.setName("Fuji Apple");
        fujiApple.setCost("1.99");
        allFoods.add(fujiApple);
    }

    private void fetchFoodTypes() {
        IFoodDAO foodDAO = RetrofitClientInstance.getRetrofitInstance().create(IFoodDAO.class);
        Call<List<FoodType>> call = foodDAO.getFoodTypes();
        //Call<List<FoodType>> call = foodDAO.getFoodTypes();
        call.enqueue(new Callback<List<FoodType>>() {
            @Override
            public void onResponse(Call<List<FoodType>> call, Response<List<FoodType>> response) {
                List<FoodType> body = response.body();
                ArrayAdapter<FoodType> adapter = new ArrayAdapter<FoodType>(MainActivity.this, android.R.layout.simple_list_item_1, body);
                actType.setAdapter(adapter);

            }

            @Override
            public void onFailure(Call<List<FoodType>> call, Throwable t) {

            }
        });
    }

    private void updateSpinner() {
        ArrayAdapter<Food> adapter = new ArrayAdapter<Food>(this, R.layout.support_simple_spinner_dropdown_item, allFoods);
        spnFood.setAdapter(adapter);
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.menu_main, menu);
        return true;
    }

    public void save(View v) {
        food.setName(edtName.getText().toString());
        food.setCost(edtCost.getText().toString());
        food.setPrepTime(edtPrepTime.getText().toString());
        food.setType(actType.getText().toString());
        int tempCalories = 0;
        if (edtCalories.getText().toString().trim().length() > 0) {
            tempCalories = Integer.parseInt(edtCalories.getText().toString());
        }
        food.setCalories(tempCalories);
        allFoods.add(food);
        food = new Food();
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        int id = item.getItemId();

        if (id == R.id.action_settings) {
            return true;
        }

        return super.onOptionsItemSelected(item);
    }

    private void newFood() {
        food = new Food();
        actType.setText("");
        edtCalories.setText("");
        edtCost.setText("");
        edtName.setText("");
        edtPrepTime.setText("");

    }

}

File: Kotlin/KotlinSimplicity/app/src/main/java/com/dodemy/kotlinsimplicity/MainActivity.kt
package com.dodemy.kotlinsimplicity

import android.os.Bundle
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.widget.AdapterView
import android.widget.AdapterView.OnItemSelectedListener
import android.widget.ArrayAdapter
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import com.dodemy.kotlinsimplicity.dao.IFoodDAO
import com.dodemy.kotlinsimplicity.dto.Food
import com.dodemy.kotlinsimplicity.dto.FoodType
import kotlinx.android.synthetic.main.activity_main.*
import kotlinx.android.synthetic.main.content_main.*
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.async
import kotlinx.coroutines.launch

class MainActivity : AppCompatActivity() {

    private var allFoods  = ArrayList<Food>()
    private var food = Food()
    private var _foodType: List<FoodType>? = ArrayList<FoodType>()

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        setSupportActionBar(toolbar)

        fab.setOnClickListener { view ->
            newFood()
        }

        actType.setOnItemClickListener { parent, view, position, id ->
            var selectedFoodType = parent.getItemAtPosition(position) as FoodType
            var foodTypeName = selectedFoodType.type
            Toast.makeText(this, " $foodTypeName  Very nutritious!", Toast.LENGTH_LONG).show()
        }

        btnSave.setOnClickListener {
            save()
        }

        spnFood.onItemSelectedListener = object : OnItemSelectedListener {
            override fun onNothingSelected(parent: AdapterView<*>?) {
                TODO("not implemented") //To change body of created functions use File | Settings | File Templates.
            }

            override fun onItemSelected(
                parent: AdapterView<*>?,
                view: View?,
                position: Int,
                id: Long
            ) {
                food = parent?.getItemAtPosition(position) as Food
                with(food) {
                    actType.setText(type)
                    edtCalories.setText(calories.toString())
                    edtCost.setText(cost)
                    edtName.setText(name.properCase())
                    edtPrepTime.setText(prepTime)
                }
            }

        }
        prepopulateFoods()
        updateSpinner()
        fetchFoodTypes()

    }

    private fun newFood() {
        food = Food();
        actType.setText("")
        edtCalories.setText("")
        edtCost.setText("")
        edtName.setText("")
        edtPrepTime.setText("")
    }

    private fun prepopulateFoods() {
        val mustard = Food("Mustard", 0, "1.29", "0")
        allFoods.add(mustard)
        val pickles = Food("Pickles")
        allFoods.add(pickles)
        val pawpaw = Food("Paw Paw", 100, "10.00")
        allFoods.add(pawpaw)
        val fujiApple = Food(name="Fuji Apple", cost="1.99")
        allFoods.add(fujiApple)
    }

    private fun save() {
        with (food) {
            name = edtName.text.toString()
            calories = if (edtCalories.text.toString().trim().isNotEmpty()) {
                Integer.parseInt(edtCalories.text.toString())
            } else {
                0
            }
            cost = edtCost.text.toString()
            prepTime = edtPrepTime.text.toString()
            type = actType.text.toString()
        }
        allFoods.add(food)
        food = Food()
    }

    private fun updateSpinner() {
        val adapter = ArrayAdapter(this, R.layout.support_simple_spinner_dropdown_item, allFoods)
        spnFood.adapter = adapter
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.menu_main, menu)
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        return when (item.itemId) {
            R.id.action_settings -> true
            else -> super.onOptionsItemSelected(item)
        }
    }

    private fun fetchFoodTypes() {

        var scope = CoroutineScope(Dispatchers.Main)
            scope.launch {
                val service = RetrofitClientInstance.retrofitInstance?.create(IFoodDAO::class.java)
                val receivedFoodType = async { service?.getFoodTypes() }
                _foodType =  receivedFoodType.await()
                val adapter = ArrayAdapter(this@MainActivity, R.layout.support_simple_spinner_dropdown_item, _foodType!!)
                actType.setAdapter(adapter)
            }

    }

    fun String.properCase() : String {
        return if (this.length > 1) {
            this.substring(0,1).toUpperCase() + this.substring(1).toLowerCase()
        } else {
            this
        }
    }
}
